#VRML_SIM R2025a utf8

WorldInfo {
  info [
    "S4 Remote Robot Management System - Omnidirectional Robot"
    "Fitfest25 Hackathon"
  ]
  title "S4 Robot World - Omni Wheels"
  basicTimeStep 64
  contactProperties [
    ContactProperties {
      material1 "wheel"
      material2 "floor"
      coulombFriction [
        0.5
      ]
      bounce 0
      softCFM 0.0001
    }
  ]
}

Viewpoint {
  orientation -0.5 0.5 0.7 2.0
  position 0 0 10
}

Background {
  skyColor [
    0.4 0.7 1
  ]
}

DirectionalLight {
  direction 0 -1 -0.5
  castShadows TRUE
}

DEF FLOOR Solid {
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.8 0.8
        roughness 1
        metalness 0
      }
      geometry Plane {
        size 20 20
      }
    }
  ]
  name "floor"
  contactMaterial "floor"
  boundingObject Plane {
    size 20 20
  }
}

DEF ROBOT Robot {
  translation 0 0 0.055
  rotation 0 0 1 0
  children [
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.2 0.4 0.8
        metalness 0
      }
      geometry Box {
        size 0.2 0.2 0.06
      }
    }
    DEF FRONT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0.1 0 0
      }
      device [
        RotationalMotor {
          name "front wheel motor"
          maxVelocity 10
          maxTorque 5.0
        }
      ]
      endPoint Solid {
        translation 0.1 0 0
        rotation 1 0 0 1.5708
        children [
          DEF WHEEL_SHAPE Shape {
            appearance PBRAppearance {
              baseColor 0.1 0.1 0.1
              roughness 1
            }
            geometry Cylinder {
              height 0.03
              radius 0.04
            }
          }
        ]
        name "front wheel"
        contactMaterial "wheel"
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density -1
          mass 0.1
        }
      }
    }
    DEF BACK_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.1 0 0
      }
      device [
        RotationalMotor {
          name "back wheel motor"
          maxVelocity 10
          maxTorque 5.0
        }
      ]
      endPoint Solid {
        translation -0.1 0 0
        rotation 1 0 0 1.5708
        children [
          USE WHEEL_SHAPE
        ]
        name "back wheel"
        contactMaterial "wheel"
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density -1
          mass 0.1
        }
      }
    }
    DEF LEFT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 1 0 0
        anchor 0 0.1 0
      }
      device [
        RotationalMotor {
          name "left wheel motor"
          maxVelocity 10
          maxTorque 5.0
        }
      ]
      endPoint Solid {
        translation 0 0.1 0
        rotation 0 1 0 1.5708
        children [
          USE WHEEL_SHAPE
        ]
        name "left wheel"
        contactMaterial "wheel"
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density -1
          mass 0.1
        }
      }
    }
    DEF RIGHT_WHEEL HingeJoint {
      jointParameters HingeJointParameters {
        axis 1 0 0
        anchor 0 -0.1 0
      }
      device [
        RotationalMotor {
          name "right wheel motor"
          maxVelocity 10
          maxTorque 5.0
        }
      ]
      endPoint Solid {
        translation 0 -0.1 0
        rotation 0 1 0 1.5708
        children [
          USE WHEEL_SHAPE
        ]
        name "right wheel"
        contactMaterial "wheel"
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density -1
          mass 0.1
        }
      }
    }
    GPS {
      name "gps"
    }
    Compass {
      name "compass"
    }
  ]
  name "robot"
  boundingObject USE BODY
  physics Physics {
    density -1
    mass 0.3
    centerOfMass [
      0 0 0
    ]
  }
  controller "robot_controller"
}
